src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1575,19 +1577,20 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceOrExtensionDeclaratio
             var name = this.ParseIdentifierToken();
             var typeParameters = this.ParseTypeParameterList();
 
-            var paramList = CurrentToken.Kind == SyntaxKind.OpenParenToken && mainKeyword.Kind != SyntaxKind.ExtensionKeyword // PROTOTYPE
-                ? ParseParenthesizedParameterList() : null;
+            var paramList = CurrentToken.Kind == SyntaxKind.OpenParenToken
+                ? ParseParenthesizedParameterList()
+                : null;
 
-            SyntaxToken? forKeyword = null;
-            TypeSyntax? forType = null;
-            if (mainKeyword.Kind == SyntaxKind.ExtensionKeyword
-                && CurrentToken.Kind == SyntaxKind.ForKeyword)
+            if (paramList != null && mainKeyword.Kind == SyntaxKind.ExtensionKeyword)
             {
-                // PROTOTYPE consider error recovery for `class X for type`
-                forKeyword = EatToken(SyntaxKind.ForKeyword);
-                forType = ParseType();
+                paramList = AddError(paramList, ErrorCode.ERR_ExtensionPrimaryConstructor);
```

Comment:
>paramList = AddError(paramList, ErrorCode.ERR_ExtensionPrimaryConstructor); [](http://example.com/codeflow?start=16&length=75)

I an somewhat surprised to see that we report this error in the parser. I thought our current approach is to delay error reporting to later phases whenever possible.

Reply:
It looks like for interface we check that somewhere else and we should detect extension types there as well. Also, it looks like we report `ERR_UnexpectedParameterList` for interfaces, I think that error is good enough for extension types as well. Consider removing the new error.

Reply:
sounds good.